Documents/UpgradeExtensibility.md

```diff
@@ -0,0 +1,157 @@
+    			     installers are run.
+
+This means the top level structure of the NuGet package will contain
+the following directories:
+
+  G4W :: Contains Git for windows installer
+  
+  GVFS :: Contains the VFSForGitInstaller
+  
+  PreSetup :: (Optional) Contains the installer to run before setup
+  
+  PostSetup :: (Optional) Contains the installer to run after setup
+
+#### Alternative 1
+
+Instead of the package containing 3 different installers, it could
+contain a single installer that would be responsible for upgrading
+Git, GVFS, and any other components. This would be simpler from the
+VFSForGit Upgrade logic, as it would only need to run a single
+installer, but the installer generation would be more complex, as it
+would need to include logic for all updated packaged - including the
+git and VFSForGit installers. As we already have the logic to run the
+git and VFSForGit installers, we decided against this option because
+it adds more complexity to the installer generation.
+
+#### Alternative 2
+
+Git and GVFS would be upgraded from the public feed as normal, and
+only the extra components would be installed from the specified
+feed. We decided against this because the components to be distributed
+can be coupled, and having two independent feeds could make it
+difficult to keep these in sync.
+
+### Upgrade Logic
+
+There are currently two high level components in VFSForGit:
+
+UpgradeOrchestrator :: Responsible for orchestrating the upgrade steps
+
+ProductUpgrader :: Responsible for details of interacting with GitHub
+Releases API, downloading assets, and performing the upgrade of
+individual assets.
+
+We will make the responsibilities of these components more clear, such
+that the UpgradeOrchestrator will be responsible for the high level
+general steps of an upgrade, while the ProductUpgrader will contain
+the logic for interacting with the different feeds and installing
+packages from these feeds. In this model, we would have a
+"GitHubReleasesUpgrader" and a "NuGetPackageUpgrader"
+
+The UpgradeOrchestrator will delegate the actual upgrade logic to
+individual ProductUpgraders. The existing upgrade logic for upgrading
+from the public GitHub releases API will be one ProductUpgrader
+implementation. We will implement another ProductUpgrader for the
+upgrades from NuGet feeds.
+
+- Checking for a new version
+  - Delegate the logic of querying the feed to Upgrader
+- Running prechecks
+- Downloading new version
+  - Download logic delegated to ProductUpgrader
+- Unmounting all repositories
+- Running installer
+  - Installer logic delegated to ProductUpgrader
+- Remounting repositories
+- Deleting downloaded files
+  - cleanup delegated to Upgrader
+
+#### Upgrade with Alternative feed
+
+When upgrading from NuGet package feed, the NuGet package will be
+downloaded and the individual well know directories extracted. The
+installers from these directories will then be run.

```

Comment:
Nugent packages generally aren't downloaded and then searched for particular files to run.  I do ensure the file names don't changes and this is extensible going forward.